Add preferences dialog

k-dominik opened the pull request:
This PR makes the settings, so far governed only by `ilastik.ini`, accessible to users via the UI:

<img width="612" alt="ilastik preferences dialog" src="https://github.com/user-attachments/assets/2f4a6e22-cabc-4bfc-9f49-07c600e50fd8">


In the process I introduced a config class based on `pydantic`. The settings, edited via the dialog, will be written to `ilastik.ini` - this is fully compatible with the settings file so far. Note: Only settings different to default values are written to the file.

Currently a full config would look like this:

<details><summary>Example <code>ilastik.ini</code></summary>

```
[ilastik]
plugin_directories = ~/my_plugins
output_filename_format = {dataset_dir}/{nickname}_output_{result_type}
output_format = tiff
logging_config = /path/to/some/logging_config.json
debug = True
hbp = True

[lazyflow]
threads = 50
total_ram_mb = 10000
```

</details>


Todos:
- [x] reading/writing of config
- [x] user-friendly tooltips/validation
- [x] investigate `pydantic` (`pydantic-settings`, maybe) 🤡 
   in the end `pydantic-settings` wasn't worth it - we don't want most of it's features (no need for .env files, we don't want to expose every setting via env variables, and also not via argparse (let alone that all of these would need to be customized to integrate with our current "interface") 


<!-- Thank you very much for opening a PR!

     Please summarize your your pull request in 1-2 sentences. -->

## Checklist

<!-- Checking off an item means that an action has been successfully completed.
     Some items might not be applicable - you can remove those if you decide
     that this is the case.
-->

- [x] Format code and imports.
- [x] Add docstrings and comments.
- [x] Add tests.
- [x] Update [user documentation](https://github.com/ilastik/ilastik.github.io). -> https://github.com/ilastik/ilastik.github.io/pull/278
- [x] Reference relevant issues and other pull requests.
- [x] Rebase commits into a logical sequence.

Fixes #2863

k-dominik committed:
make default config path to always have a default

k-dominik committed:
Pydantic-based ilastik settings

k-dominik committed:
Added basic preferences dialog

k-dominik committed:
Added tests for preference dialog

btbest commented on ilastik/config.py:
```diff
@@ -63,25 +132,36 @@ class RuntimeCfg:
+    config_path = Path(path)
+
+    if config_path.is_file():
+        logger.info(f"Loading configuration from {config_path!s}.")
```
Note that this will not actually be logged unless the config file path is specified with `--configfile=` (i.e. init called through `app.main()`, not during a regular load from ilastik.ini in the default location.

I think that's fine (and not a consequence of this PR anyway), but the import-time init in this module isn't a great choice for tracing what init happens when and where.

btbest reviewed (commented):
Edit: This was meant to be "Request changes" because `preferencesDialog.py` has a bug in an else-raise-clause due to variable naming, `default_config` should be deleted (I think) and the `flat_dict` should be explained or removed.

Really nice, the widget generator and the prefs based on pydantic will make it pretty easy to extend the configurable options in the future.

I see a potential for confusion because we have a `preferences.json`, which the user can locate with `Open Config Folder` - but these are the volumina preferences. Now we have a Preferences dialog that controls a different set of preferences that we internally call config.

From the dev perspective, it might be worth just adding a short note to the respective docstrings to explain that ilastik-config goes with PreferencesDialog and that these are separate from `preferences`, which is from volumina. Though the code is easy enough to trace anyway.

From the user perspective, I'm not really sure what can be done. Maybe creating the ilastik.ini file with the default values could help, so that at least if a user uses Open Config Folder, they can see that there are in fact two separate sets of settings?

k-dominik commented on ilastik/config.py:
```diff
@@ -63,25 +132,36 @@ class RuntimeCfg:
-    config.read_string(default_config)
-    if config_path is not None:
-        config.read(config_path)
+        flat_dict = {}
```
I could swear that it was necessary to flatten the dict - but I tried now passing `_cfg` directly and it just works™️. Will remove.

k-dominik committed:
Review comments/changes

* removed flattening of config dict - just works without
* removed `default_config` (not used) -> moved information docstring as example
* removed qtbot fixture from test that don't need it
* renamed `Increment.increment` member to `inc` to avoid confusion

Co-authored-by: Benedikt Best <63287233+btbest@users.noreply.github.com>

k-dominik commented on ilastik/config.py:
```diff
@@ -63,25 +132,36 @@ class RuntimeCfg:
-        config.read(config_path)
+        flat_dict = {}
+        for k, v in _cfg.items():
+            if not isinstance(v, Mapping):
+                flat_dict[k] = v
+            else:
+                flat_dict[k] = {sk: sv for sk, sv in v.items()}
```
code removed

k-dominik commented on ilastik/config.py:
```diff
@@ -44,6 +49,70 @@
+
+@dataclass
+class Increment:
+    increment: int
```
:D

k-dominik committed:
Strip whitespace from string options

Co-authored-by: Benedikt Best <63287233+btbest@users.noreply.github.com>

k-dominik committed:
Add advanced options - hidden by default

Some options are not validated very well. For now we hide those to
prevent changing them without much consideration.

k-dominik commented on ilastik/config.py:
```diff
@@ -44,6 +49,70 @@
+
+class IlastikSection(_ConfigBase):
+
+    plugin_directories: str = "~/.ilastik/plugins,"
```
your comment sent me into a "Sinnkrise" -- the first step to overengineering ;)

Currently validation on all of the strings is very nimble. Especially for the comma-separated list of paths - where `,` is a valid character for Paths on all OSses... So a "strip whitespace" validator will not do the job. This does not change with this PR, but it _does_ make it more accessible and hence more prone to cause problems.

without major work I see 3 options

1) scrap this
2) keep as is, and accept that changing the string values here can lead to all kinds of errors.
3) hide at least some of the options that we don't want to expose to all users

I'm leaning towards 3)... which I implemented in https://github.com/ilastik/ilastik/pull/2899/commits/5bc90554eaf906f1a8d537ead919171432b5d569 (also happy to remove if too controversial)

k-dominik replied:
what do you think @btbest? :)

btbest replied:
Sorry, I got bogged down in details instead of maybe just replying... I also like 3 the most - especially as advanced options that are still available if the user really looks for them, as you've implemented now :)

k-dominik committed:
Merge branch 'main' into settings-dlg

btbest reviewed (changes requested):
:/

I realised there's an issue with coupling the dialog's model directly to the in-use runtime config: Regardless of whether you Ok or Cancel the dialog (which governs whether you write to disk), any non-default values you entered in the inputs are kept in the active runtime config. Also, if a non-default value was read from file, then resetting it to default in the dialog and clicking Ok doesn't actually transfer the value to the model.

The dialog does tell the user to please restart ilastik, but that should only be a precaution and not a requirement in order for Ok and Cancel to modify (or not modify) the ongoing runtime configs...

Request changes because there's also a remaining usage of the old config dict syntax in dataExportGui: `settingsDlg = DataExportOptionsDlg(self, opExportModelOp, cfg["ilastik"]["output_filename_format"])`
